src/Database/LSMTree/Internal/WriteBufferWriter.hs

```diff
@@ -0,0 +1,227 @@
+     (MonadST m, MonadSTM m, MonadThrow m)
+  => Bool -- ^ drop caches
+  -> WriteBufferWriter m h
+  -> m (HasFS m h, HasBlockIO m h, WriteBufferFsPaths)
+unsafeFinalise dropCaches WriteBufferWriter {..} = do
+  -- write final bits
+  mPage <- ST.stToIO $ flushPageIfNonEmpty writeBufferWriterPageAcc
+  for_ mPage $ writeRawPage writeBufferWriterHasFS (forWriteBufferKOps writeBufferWriterHandles)
+  checksums <- toChecksumsFileForWriteBufferFiles <$> traverse readChecksum writeBufferWriterHandles
+  FS.withFile writeBufferWriterHasFS (writeBufferChecksumsPath writeBufferWriterFsPaths) (FS.WriteMode FS.MustBeNew) $ \h -> do
+    CRC.writeChecksumsFile' writeBufferWriterHasFS h checksums
+    FS.hDropCacheAll writeBufferWriterHasBlockIO h
+  -- drop the KOps and blobs files from the cache if asked for
+  when dropCaches $ do
+    dropCache writeBufferWriterHasBlockIO (forWriteBufferKOpsRaw writeBufferWriterHandles)
+    dropCache writeBufferWriterHasBlockIO (forWriteBufferBlobRaw writeBufferWriterHandles)
+  for_ writeBufferWriterHandles $ closeHandle writeBufferWriterHasFS
+  return (writeBufferWriterHasFS, writeBufferWriterHasBlockIO, writeBufferWriterFsPaths)
+
+
+-- | See 'RunBuilder.addKeyOp'.
+addKeyOp ::
+     (MonadST m, MonadSTM m, MonadThrow m)
+  => WriteBufferWriter m h
+  -> SerialisedKey
+  -> Entry SerialisedValue (RawBlobRef m h)
+  -> m ()
+addKeyOp WriteBufferWriter{..} key op = do
+  op' <- traverse (copyBlob writeBufferWriterHasFS writeBufferWriterBlobOffset (forWriteBufferBlob writeBufferWriterHandles)) op
+  if PageAcc.entryWouldFitInPage key op
+    then do
+      mPage <- ST.stToIO $ addSmallKeyOp writeBufferWriterPageAcc key op'
+      for_ mPage $ writeRawPage writeBufferWriterHasFS (forWriteBufferKOps writeBufferWriterHandles)
+    else do
+      (pages, overflowPages) <- ST.stToIO $ addLargeKeyOp writeBufferWriterPageAcc key op'
+      --TODO: consider optimisation: use writev to write all pages in one go (see RunBuilder)
+      for_ pages $ writeRawPage writeBufferWriterHasFS (forWriteBufferKOps writeBufferWriterHandles)
+      writeRawOverflowPages writeBufferWriterHasFS (forWriteBufferKOps writeBufferWriterHandles) overflowPages
+
+-- addLargeSerialisedKeyOp ::
+--      (MonadST m, MonadSTM m)
+--   => WriteBufferWriter m h
+--   -> RawPage
+--   -> [RawOverflowPage]
+--   -> m ()
+-- addLargeSerialisedKeyOp WriteBufferWriter{..} page overflowPages =
+--   assert (RawPage.rawPageNumKeys page == 1) $
+--   assert (RawPage.rawPageHasBlobSpanAt page 0 == 0) $
+--   assert (RawPage.rawPageOverflowPages page > 0) $
+--   assert (RawPage.rawPageOverflowPages page == length overflowPages) $ do
+--     !pages <- ST.stToIO $ selectPages <$> flushPageIfNonEmpty writeBufferWriterPageAcc <*> pure page
+--     for_ pages $ writeRawPage writeBufferWriterHasFS (forWriteBufferKOps writeBufferWriterHandles)
+--     writeRawOverflowPages writeBufferWriterHasFS (forWriteBufferKOps writeBufferWriterHandles) overflowPages
```

Comment:
If it is not used, we can remove the code

```suggestion
```

Reply:
It's largely in there just in case we decide to export the incremental API, in which case it might be needed.

Reply:
It's largely in there just in case we decide to export the incremental API, in which case it might be needed.